test/remote/gateways/remote_shift4_test.rb

```diff
@@ -0,0 +1,321 @@
+      },
+      "card": {},
+      access_token: '4902FAD2-E88F-4A8D-98C2-EED2A73DBBE2'
+    }
+    @purchase_params = {
+      company_name: 'Spreedly',
+      "api_options": [
+        'ALLOWPARTIALAUTH'
+      ],
+      "amount": {
+        "tax": 0,
+        "total": 219
+      },
+      "clerk": {
+        "numeric_id": 16
+      },
+      "transaction": {
+        "invoice": '4666309473',
+        "purchase_card": {
+          "customer_reference": '1234567',
+          "destination_postal_code": '89123',
+          "product_descriptors": 'Test'
+        },
+        "notes": 'Test payment'
+      },
+      "card": {
+        "expiration_date": '0825',
+        "present": 'N',
+        "indicator": '1',
+        "value": '3333'
+      },
+      access_token: '4902FAD2-E88F-4A8D-98C2-EED2A73DBBE2'
+    }
+
+    @access_token =
+    @invoice = '3333333309'
+    @token = '8042677003331111'
+    @void_options = { company_name: 'Spreedly', invoice: @invoice, access_token: '4902FAD2-E88F-4A8D-98C2-EED2A73DBBE2' }
+  end
+
+  def test_successful_access_token
+    response = @gateway.access_token(@access_token_opt)
+    @access_token = credentials(response)['accessToken']
+
+    assert response.success?
+    assert_equal success_message(response), 'Transaction successful'
+    assert_equal access_token_success_from(response), true
+  end
+
+  def test_successful_authorize
+    response = @gateway.authorize(@amount, @credit_card, @options)
+
+    assert response.success?
+    assert_equal success_message(response), 'Transaction successful'
+    assert_equal @options[:amount][:total], verify_amount(response)
+  end
+
+  def test_successful_capture
+    authorize_res = @gateway.authorize(@amount, @credit_card, @options)
+    auth_token = authorize_res.authorization
+
+    response = @gateway.capture(@amount, auth_token, @capture_params)
+    assert success?(response)
+
+    assert response.success?
+    assert_equal success_message(response), 'Transaction successful'
+    assert_equal @options[:amount][:total], verify_amount(response)
+    assert response_result(response)['transaction']['invoice'].present?
+  end
+
+  def test_successful_purchase
+    auth_token = '8042677003331111'
+    response = @gateway.purchase(@amount, auth_token, @purchase_params)
+    assert success?(response)
+
+    assert_equal @options[:amount][:total], verify_amount(response)
+    assert successful_transaction?(response)
+  end
+
+  def test_failed_purchase
+    auth_token = '8038483489222221'
+    @purchase_params[:card][:expiration_date] = '085'
+    response = @gateway.purchase(@amount, auth_token, @purchase_params)
+
+    assert error?(response)
+    assert error_msg(response)
+  end
+
+  def test_failed_authorize
+    response = @gateway.authorize(@amount, @credit_card, @options.except(:access_token))
+
+    assert error?(response)
+    assert merchant_error(response)
+  end
+
+  def test_failed_capture
+    auth_token = '8038483489222221'
```

Comment:
Since both `auth_tokens` are used twice I think they should be put into a variable.

Also, minor suggestion you can take or leave but I think `card_token` might be a better name for the variable.

Do you happen to know if the token is then usable directly with the Shift4 API? Are we vaulting the PMs at Shift4? I'm wondering if this way of passing a token instead of PAN data is an example of using a card that's been `stored`.

Reply:
@drkjc Both have different values you can verify them as well.

Its actually card auth token and its being used as authorisation token that why named it auth token. @drkjc

Reply:
@ali-hassan Gotcha, that makes sense. Thanks. I'll resolve this one.